packages/components/src/presentation-components/viewport/ViewportSelectionHandler.ts

```diff
@@ -69,54 +63,42 @@ export class ViewportSelectionHandler implements IDisposable {
     this._imodel.hilited.wantSyncWithSelectionSet = false;
     this._selectionHandler.imodel = value;
 
-    void this.applyCurrentSelection();
+    this.applyCurrentSelection();
   }
 
-  /** note: used only it tests */
-  public get pendingAsyncs() {
-    return this._asyncsTracker.pendingAsyncs;
+  public applyCurrentSelection() {
+    this.applyUnifiedSelection(this._imodel);
   }
 
-  public async applyCurrentSelection() {
-    await this.applyUnifiedSelection(this._imodel, { providerName: "" }, this.selectionHandler.getSelection());
-  }
+  private applyUnifiedSelection(imodel: IModelConnection) {
+    this._cancelOngoingChanges.next();
 
-  private async applyUnifiedSelection(imodel: IModelConnection, selectionInfo: SelectionInfo, selection: Readonly<KeySet>) {
-    if (this._asyncsTracker.pendingAsyncs.size > 0) {
-      this._lastPendingSelectionChange = { info: selectionInfo, selection };
-      return;
-    }
-
-    await using(this._asyncsTracker.trackAsyncTask(), async (_r) => {
-      const ids = await Presentation.selection.getHiliteSet(this._imodel);
-      using(Presentation.selection.suspendIModelToolSelectionSync(this._imodel), (_) => {
-        imodel.hilited.clear();
-        let shouldClearSelectionSet = true;
-        if (ids.models && ids.models.length) {
-          imodel.hilited.models.addIds(ids.models);
-        }
-        if (ids.subCategories && ids.subCategories.length) {
-          imodel.hilited.subcategories.addIds(ids.subCategories);
-        }
-        if (ids.elements && ids.elements.length) {
-          imodel.hilited.elements.addIds(ids.elements);
-          imodel.selectionSet.replace(ids.elements);
-          shouldClearSelectionSet = false;
-        }
-        if (shouldClearSelectionSet) {
-          imodel.selectionSet.emptyAll();
-        }
-      });
+    using(Presentation.selection.suspendIModelToolSelectionSync(this._imodel), (_) => {
+      imodel.hilited.clear();
```

Comment:
@grigasp What do you think about clearing hilited elements before requesting new hilite set? If you don't like this approach I will update `getHiliteSetIterator` implementation to emit empty hilite set if there are node keys.

Reply:
I think we have to consider the situation where user picks an element in the viewport. Here's what happens:
1. User picks an element.
2. The element is added to viewport's selection and hilite sets. The element gets hilited.
3. Unified selection is updated with the picked element (let's ignore selection scopes here).
4. We request a hilite set for the new selection, from now on - in batches.
5. On each batch, we apply it on viewport's selection and hilite sets to get element displayed.

For the user, steps 3-4 are invisible, so all he sees after picking an element is it getting hilited immediately in step 2 and, in cases when we get a larger hilite set, more elements getting hilited in a few moments. 

If we clear viewport's selection/hilite sets between steps 3 and 4, the user would see the picked element hilite, then de-hilite, and then hilite again. I don't think we want that.

I see the previous implementation would first get the full hilite set and then clear before applying the new one. With batched requests this isn't as simple as we want to update the viewport with new batch as soon as we receive it, but it isn't guaranteed that the first batch contains all initially picked elements (e.g. when user uses fence selection there could be more elements that we have in a single batch). However, I think we could say that it's guaranteed for initially picked elements to be a subset of the final hilite set. Maybe we don't have to clear anything at all then?

Reply:
As discussed in private. Changed implementation to clear hilite set only when selection change event is of type `Replace` or `Clear`. For `Add` and `Remove` current implementation add/removes only affected elements to/from hilite set.